packages/oauth/oauth-provider/src/router/api-router.ts

```diff
@@ -0,0 +1,899 @@
+    M extends 'GET' | 'POST',
+    E extends `/${string}` &
+      // Extract all the endpoint path that match the method (allows for
+      // auto-complete & better error reporting)
+      {
+        [E in keyof ApiEndpoints]: ApiEndpoints[E] extends { method: M }
+          ? E
+          : never
+      }[keyof ApiEndpoints],
+    S extends // A schema that validates the POST input or GET params
+      ApiEndpoints[E] extends { method: 'POST'; input: infer I }
+        ? z.ZodType<I>
+        : ApiEndpoints[E] extends { method: 'GET'; params: infer P }
+          ? z.ZodType<P>
+          : void,
+  >(
+    method: M,
+    endpoint: E,
+    schema: S,
+    buildJson: (
+      this: ApiContext<RouteCtx<C>, InferValidation<S>>,
+      req: TReq,
+      res: TRes,
+    ) => Awaitable<ApiEndpoints[E]['output']>,
+  ): Middleware<C, TReq, TRes> {
+    return createRoute(
+      method,
+      `${API_ENDPOINT_PREFIX}${endpoint}`,
+      apiMiddleware(method, schema, buildJson),
+    )
+  }
+
+  function apiMiddleware<C extends RouterCtx, S extends void | z.ZodTypeAny>(
+    method: 'GET' | 'POST',
+    schema: S,
+    buildJson: (
+      this: ApiContext<C, InferValidation<S>>,
+      req: TReq,
+      res: TRes,
+    ) => unknown,
+  ): Middleware<C, TReq, TRes> {
+    const parseInput: (this: C, req: TReq) => Promise<InferValidation<S>> =
+      schema == null // No schema means endpoint doesn't accept any input
+        ? async function (req) {
+            req.resume() // Flush body
+            return undefined
+          }
+        : method === 'POST'
+          ? async function (req) {
+              const body = await parseHttpRequest(req, ['json'])
+              return schema.parseAsync(body, { path: ['body'] })
+            }
+          : async function (req) {
+              // @NOTE This should not be necessary with GET requests
+              req.resume().once('error', (_err) => {
+                // Ignore errors when flushing the request body
+                // (e.g. client closed connection)
+              })
+
+              const query = Object.fromEntries(this.url.searchParams)
+              return schema.parseAsync(query, { path: ['query'] })
+            }
+
+    return jsonHandler<C, TReq, TRes>(async function (req, res) {
+      try {
+        // Prevent caching of API routes
+        res.setHeader('Cache-Control', 'no-store')
+        res.setHeader('Pragma', 'no-cache')
+
+        // Prevent CORS requests
+        validateFetchMode(req, ['same-origin'])
+        validateFetchSite(req, ['same-origin'])
+        validateOrigin(req, issuerOrigin)
+        const referer = validateReferer(req, { origin: issuerOrigin })
+
+        // Ensure we are one the right page
+        if (
+          referer.pathname !== '/oauth/authorize' &&
+          referer.pathname !== '/account' &&
+          !referer.pathname.startsWith(`/account/`)
+        ) {
+          throw createHttpError(400, `Invalid referer ${referer}`)
+        }
```

Comment:
I am not sure if trailing slashes are allowed on those pages, but if they are we might consider permitting them here.

Reply:
We actually don't. Added a comment.